### PR TITLE
LPS-142654 Add permissions to AccountRoles

### PR DESCRIPTION
**Pull request by pei-jung**

### CODE REVIEW

**Comment by pei-jung** [comment]
ci:test:accounts

**Comment by pei-jung** [comment]
ci:test:account

**Comment by epark18** [comment]
Hello, @pei-jung there is one unique failure in relevant for PortalSmokeUpgrade#ViewPortalSmokeArchive7413. [Logs](https://testray.liferay.com/reports/production/logs/2021-12/test-1-27/test-portal-acceptance-pullrequest(master)/954/functional-upgrade-tomcat90-mysql57-jdk8/0/2/jenkins-console.txt.gz) | [Case Result](https://testray.liferay.com/home/-/testray/case_results/1587084998)
`     [java] 2021-12-03 08:22:52.175 INFO  [main][UpgradeProcess:134] Failed upgrade process com.liferay.account.internal.upgrade.v2_5_0.AccountRoleResourceUpgradeProcess in 62 ms
     [java] 2021-12-03 08:22:55.440 ERROR [main][UpgradeStepRegistratorTracker:143] Failed upgrade process for module com.liferay.account.service
     [java] com.liferay.portal.kernel.upgrade.UpgradeException: java.sql.SQLIntegrityConstraintViolationException: Column 'companyId' in field list is ambiguous
     [java] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:128) ~[portal-kernel.jar:?]`

**Comment by epark18** [comment]
Most unique failures stem from the same issue with Accounts being unavailable whenever trying to add an account through the Accounts Admin. [Example Case Result](https://testray.liferay.com/home/-/testray/case_results/1587293439) | [LOGS](https://testray.liferay.com/reports/production/logs/2021-12/test-1-26/test-portal-acceptance-pullrequest(master)/971/functional-tomcat90-mysql57-jdk8/0/20/jenkins-console.txt.gz) | [POSHI Report](https://testray.liferay.com/reports/production/logs/2021-12/test-1-26/test-portal-acceptance-pullrequest(master)/971/functional-tomcat90-mysql57-jdk8/0/20/LocalFile.Accounts_Smoke/index.html.gz)
image

Some of the 2B Account Management Widget tests are failing because it looks like the "Organizations" tab is no longer listed when editing. [Case Result](https://testray.liferay.com/home/-/testray/case_results/1587293338) | [LOGS](https://testray.liferay.com/reports/production/logs/2021-12/test-1-26/test-portal-acceptance-pullrequest(master)/971/functional-tomcat90-mysql57-jdk8/0/1/jenkins-console.txt.gz) | [POSHI Report](https://testray.liferay.com/reports/production/logs/2021-12/test-1-26/test-portal-acceptance-pullrequest(master)/971/functional-tomcat90-mysql57-jdk8/0/1/LocalFile.2BAccount_AccountAdminCanUnassignOrganizationsInBulk/index.html.gz)
image

`2BAccount#AccountAdminCanAssignRolesToUsers` This test looks like it's failing because it is navigating back to the Accounts table after assigning a role to a user. I think normal tendency was for the page to stay at Account Users after assigning so that's likely why the assert is failing here. [Case Result](https://testray.liferay.com/home/-/testray/case_results/1587293366?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D1587293249%26testrayRunId%3D1587293278%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc) | [LOGS ](https://testray.liferay.com/reports/production/logs/2021-12/test-1-26/test-portal-acceptance-pullrequest(master)/971/functional-tomcat90-mysql57-jdk8/0/0/jenkins-console.txt.gz) | [POSHI Report](https://testray.liferay.com/reports/production/logs/2021-12/test-1-26/test-portal-acceptance-pullrequest(master)/971/functional-tomcat90-mysql57-jdk8/0/0/LocalFile.2BAccount_AccountAdminCanAssignRolesToUsers/index.html.gz)

File: modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/portlet/action/SetUserAccountRolesMVCActionCommand.java [hunk 1]
```diff
@@ -70,16 +71,19 @@ protected void doProcessAction(
...
 		}
 
-		_accountRoleLocalService.associateUser(
+		_accountRoleService.associateUser(
 			accountEntryId, accountRoleIds, accountUserId);
 
 		for (AccountRole accountRole : removeAccountRoles) {
-			_accountRoleLocalService.unassociateUser(
+			_accountRoleService.unassociateUser(
 				accountEntryId, accountRole.getAccountRoleId(), accountUserId);
```
Comment (drewbrokke) [inline modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/portlet/action/SetUserAccountRolesMVCActionCommand.java:79]: For future reference: we should move this `set` operation inside a dedicated service method. It's not required right now but I figured I'd mention it.

File: modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/frontend/taglib/servlet/taglib/AccountEntryRolesScreenNavigationCategory.java [hunk 2]
```diff
@@ -69,16 +75,28 @@ public boolean isVisible(
 		if (!AllowEditAccountRoleThreadLocal.isAllowEditAccountRole() ||
 			Objects.equals(
 				accountEntryDisplay.getType(),
-				AccountConstants.ACCOUNT_ENTRY_TYPE_GUEST)) {
+				AccountConstants.ACCOUNT_ENTRY_TYPE_GUEST) ||
+			(accountEntryDisplay.getAccountEntryId() < 0)) {
```
Comment (drewbrokke) [inline modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/frontend/taglib/servlet/taglib/AccountEntryRolesScreenNavigationCategory.java:79]: What if the `accountEntryId` is `0`? Will it fail this check? I think the inverse of 
```
accountEntryDisplay.getAccountEntryId() > 0
```

would be

```
accountEntryDisplay.getAccountEntryId() <= 0
```

Am I missing something?

File: modules/apps/account/account-service/src/main/java/com/liferay/account/internal/upgrade/v2_5_0/AccountRoleResourceUpgradeProcess.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	public AccountRoleResourceUpgradeProcess(
+		ResourceLocalService resourceLocalService) {
+
+		_resourceLocalService = resourceLocalService;
+	}
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		try (PreparedStatement preparedStatement = connection.prepareStatement(
+				"select accountRoleId, companyId, userId from AccountRole " +
```
Comment (drewbrokke) [inline modules/apps/account/account-service/src/main/java/com/liferay/account/internal/upgrade/v2_5_0/AccountRoleResourceUpgradeProcess.java:38]: I think the [CI failure](https://github.com/liferay-usm/liferay-portal/pull/363#issuecomment-985938572) means you need to specify which table each column is coming from. So while the columns `accountRoleId` and `userId` are each found in only one table, both tables have a `companyId` column. So perhaps you'd just specify `AccountRole.companyId` in this case.

File: modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleServiceImpl.java [hunk 4]
```diff
@@ -49,9 +51,14 @@ public AccountRole addAccountRole(
...
-			permissionChecker, accountEntryId,
-			AccountActionKeys.ADD_ACCOUNT_ROLE);
+		if (accountEntryId > 0) {
+			_accountEntryModelResourcePermission.check(
+				permissionChecker, accountEntryId,
+				AccountActionKeys.ADD_ACCOUNT_ROLE);
+		}
+		else {
+			PortalPermissionUtil.check(permissionChecker, ActionKeys.ADD_ROLE);
+		}
```
Comment (drewbrokke) [inline modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleServiceImpl.java:61]: Looks like you fixed what my other comment mentioned :)

File: modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/security/permission/resource/AccountRolePermission.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+/**
+ * @author Pei-Jung Lan
+ */
+@Component(immediate = true, service = {})
+public class AccountRolePermission {
+
+	public static boolean contains(
+			PermissionChecker permissionChecker, AccountRole accountRole,
+			String actionId)
+		throws PortalException {
```
Comment (drewbrokke) [inline modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/security/permission/resource/AccountRolePermission.java:34]: I noticed that pretty much every usage of this class requires you to use a `try/catch` block with logging. Why not move the exception handling and logging to inside these methods, then remove the exception from these method signatures? It seems like usage would be much more straightforward then.

**Review by drewbrokke — CHANGES_REQUESTED** [review]
Nice work Pei! I have a few suggestions, look the over and tell me what you think.

File: modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleServiceImpl.java [hunk 6]
```diff
@@ -30,4 +40,97 @@
...
+	public AccountRole addAccountRole(
+			long accountEntryId, String name, Map<Locale, String> titleMap,
+			Map<Locale, String> descriptionMap)
+		throws PortalException {
+
+		PermissionChecker permissionChecker = getPermissionChecker();
+
+		_accountRoleModelResourcePermission.check(
+			permissionChecker, accountEntryId,
+			AccountActionKeys.ADD_ACCOUNT_ROLE);
```
[the remark this answers is not in the document]
Comment (drewbrokke) [reply modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleServiceImpl.java]: Nevermind, you fixed it in a later commit 👍

File: modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/frontend/taglib/servlet/taglib/AccountEntryRolesScreenNavigationCategory.java (hunk 2 above)
Comment (pei-jung) [reply modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/frontend/taglib/servlet/taglib/AccountEntryRolesScreenNavigationCategory.java:79]: Ah, you're right. It should be `accountEntryDisplay.getAccountEntryId() <= 0`. Will fix it. Thanks!

**Comment by pei-jung** [comment]
ci:test:sf

**Comment by pei-jung** [comment]
ci:test:relevant

**Comment by pei-jung** [comment]
ci:test:account

**Comment by epark18** [comment]
@pei-jung It looks like a lot of the failures are the result of a recently discovered [Regression Bug](https://issues.liferay.com/browse/LPS-143439). [LOGS](https://testray.liferay.com/reports/production/logs/2021-12/test-1-36/test-portal-acceptance-pullrequest(master)/34/modules-integration-mysql57-jdk8/0/1/jenkins-console.txt.gz) Most checks failed due to this error.

`     [exec] com.liferay.account.service.test.AccountRoleLocalServiceTest > testAddAccountRole FAILED
     [exec]     java.lang.AssertionError: Thread Thread[Framework Event Dispatcher: Equinox Container: 3fce6798-33ca-4ec3-8f04-d0b94a721b7f,5,main] caught concurrent failure: java.lang.AssertionError: {level=ERROR, loggerName=osgi.logging.Events.Framework, message=FrameworkEvent ERROR 
     [exec]     java.lang.AssertionError: {level=ERROR, loggerName=osgi.logging.Events.Framework, message=FrameworkEvent ERROR 
     [exec]     	at com.liferay.portal.test.rule.LogAssertionTestRule$LogAppender.append(LogAssertionTestRule.java:330)`

**Comment by pei-jung** [comment]
thanks @epark18! Looks like that regression bug caused a lot of the account integration tests to fail. 

Also, for the failed Accounts functional tests, I wasn't able to reproduce any of them locally. I was able to assign and unassign account roles to users in the Accounts widget. Can I get some help to reproduce the issue?

**Comment by pei-jung** [comment]
@epark18 Nevermind. I need to login as the Account Admin to assign the AccountRoles. I was able to reproduce. Will fix it and push

**Comment by pei-jung** [comment]
ci:test:relevant

**Comment by pei-jung** [comment]
ci:test:account

**Comment by pei-jung** [comment]
Hi @lr-leo, 

Do we want to give the Account Administrator the `Manage Organization` permission to assign and unassign organizations to the Account by default? Currently, the default permissions given to the Account Administrator are

- Manage Users
- Permissions
- Update
- View
- View Account Roles
- View Organizations
- View Users

**Comment by pei-jung** [comment]
@patriciajeanperez @epark18 I checked with Leo and verified that the Account Administrator doesn't have the `Manage Organizations` permission by default. This behavior is also in parity with what Commerce had originally. With that said, can you update the functional tests for the Account Administrator? I think the these are the tests that need to be updated:

- LocalFile.2BAccount#AccountAdminCanUnassignOrganizationToAccount
- LocalFile.2BAccount#AccountAdminCanUnassignOrganizationsInBulk

Thanks!

**Comment by patriciajeanperez** [comment]
ci:test:account

**Comment by patriciajeanperez** [comment]
Rerunning tests since some of them were blocked by [LPS-143439](https://issues.liferay.com/browse/LPS-143439).

**Comment by drewbrokke** [comment]
ci:forward

**Pull request closed without merging**